lib/Dialect/Ibis/Transforms/IbisContainerize.cpp

```diff
@@ -0,0 +1,134 @@
+  return uniqueName;
+}
+namespace {
+
+struct OutlineContainerPattern : public OpConversionPattern<ContainerOp> {
+  OutlineContainerPattern(MLIRContext *context, SymbolCache *symCache)
+      : OpConversionPattern<ContainerOp>(context), symCache(symCache) {}
+
+  using OpAdaptor = typename OpConversionPattern<ContainerOp>::OpAdaptor;
+
+  LogicalResult
+  matchAndRewrite(ContainerOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    // Outline the container into the module scope, by prefixing it with the
+    // parent class name.
+    auto parentClass =
+        dyn_cast_or_null<ClassOp>(op.getOperation()->getParentOp());
+    if (!parentClass)
+      return failure();
+
+    rewriter.setInsertionPoint(parentClass);
+    auto newContainerName =
+        rewriter.getStringAttr(parentClass.getName() + "_" + op.getName());
+    // unique it...
+    newContainerName =
+        getUniqueName(rewriter.getContext(), newContainerName, *symCache);
+    auto newContainer =
+        rewriter.create<ContainerOp>(op.getLoc(), newContainerName);
+    symCache->addDefinition(newContainerName, newContainer);
+
+    rewriter.mergeBlocks(op.getBodyBlock(), newContainer.getBodyBlock(), {});
+    rewriter.eraseOp(op);
+
+    // Rename the ibis.this operation to refer to the proper op.
+    auto thisOp =
+        cast<ThisOp>(cast<ScopeOpInterface>(*newContainer.getOperation())
+                         .getThis()
+                         .getDefiningOp());
+    rewriter.setInsertionPoint(thisOp);
+    rewriter.replaceOpWithNewOp<ThisOp>(thisOp, newContainer.getNameAttr());
+    return success();
+  }
+
+  SymbolCache *symCache;
+};
+
+struct ClassToContainerPattern : public OpConversionPattern<ClassOp> {
+  using OpConversionPattern<ClassOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ClassOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    // Replace the class by a container of the same name.
+    auto newContainer =
+        rewriter.create<ContainerOp>(op.getLoc(), op.getNameAttr());
+    rewriter.mergeBlocks(op.getBodyBlock(), newContainer.getBodyBlock(), {});
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+
+/// Run all the physical lowerings.
+struct ContainerizePass : public IbisContainerizeBase<ContainerizePass> {
+  void runOnOperation() override;
+
+private:
+  // Outlines containers nested within classes into the module scope.
+  LogicalResult outlineContainers();
+
+  // Converts classes to containers.
+  LogicalResult containerizeClasses();
+};
+} // anonymous namespace
+
+LogicalResult ContainerizePass::outlineContainers() {
+  auto *context = &getContext();
+  ConversionTarget target(*context);
+  target.addLegalDialect<IbisDialect>();
+  target.addDynamicallyLegalOp<ContainerOp>(
+      [&](auto *op) { return isa<mlir::ModuleOp>(op->getParentOp()); });
```

Comment:
Can you change this to "!isa<ClassOp>"? We're looking to create a different top-level op in the `hw` dialect so I want as much as possible to future-proof that in new code.